copilot/src/App/App.js

```diff
@@ -1,28 +1,129 @@
 import kind from '@enact/core/kind';
-import qs from 'query-string';
+import Popup from '@enact/agate/Popup';
+import {Cell, Row} from '@enact/ui/Layout';
+
 import React from 'react';
+import openSocket from 'socket.io-client';
 
 import css from './App.less';
 
-const args = qs.parse(typeof window !== 'undefined' ? window.location.search : '');
-
-const App = kind({
+const AppBase = kind({
 	name: 'App',
 
 	styles: {
 		css,
-		className: 'app'
+		className: 'app enact-fit'
 	},
 
-	render: (props) => {
-		const {url} = args;
+	render: ({adContent, showAd, url, popupOpen, setScreen, togglePopup, ...rest}) => {
 		return (
-			<div {...props}>
-				<iframe src={url} />
+			<div {...rest}>
+				<Popup
+					open={popupOpen}
+					closeButton
+					onClose={togglePopup}
+				>
+					<title>
+					Select Your Screen Source
+					</title>
+					<buttons>
+						<Button onClick={setScreen(1)}>Screen 1</Button>
+						<Button onClick={setScreen(2)}>Screen 2</Button>
+					</buttons>
+				</Popup>
+				<Row {...rest}>
+					<Cell
+						className={css.iframe}
+						allow="autoplay"
+						component="iframe"
+						src={url}
+					/>
+					{!showAd ? null : <Cell className={css.adSpace} shrink>
+						{adContent}
+					</Cell>}
+				</Row>
+				<Button style={{position: 'absolute'}} icon="plug" onClick={togglePopup} />
 			</div>
 		);
 	}
 });
 
+class App extends React.Component {
+	constructor (props) {
+		super(props);
+		this.state = {
+			adContent: this.props.adContent || 'Your Ad Here',
+			url: '',
+			popupOpen: true,
+			showAd: this.props.showAd || false,
+			screenId: 1
+		};
+	}
+
+	componentWillMount () {
+		this.socket = openSocket('http://localhost:3000');
+		this.listenForVideoChange(this.state.screenId);
+		this.socket.on('SHOW_AD', this.showAdSpace);
+		this.socket.emit('COPILOT_CONNECT', this.state.screenId);
+	}
+
+	componentDidUpdate (prevProps, prevState) {
+		if (prevState.screenId !== this.state.screenId) {
+			this.socket.removeAllListeners(`VIDEO_ADD_COPILOT/${prevState.screenId}`);
+			this.listenForVideoChange(this.state.screenId);
+		}
+	}
+
+	componentWillUnmount () {
+		this.socket.close();
+		clearTimeout(this.adTimer);
+	}
+
+	listenForVideoChange = (screenId) => {
+		this.socket.on(`VIDEO_ADD_COPILOT/${screenId}`, (item) => {
+			this.setState(() => {
+				return {
+					url: item.url
+				};
+			});
+		});
+	}
+
+	onToggle = () => {
+		this.setState(({popupOpen}) => {
+			return {popupOpen: !popupOpen};
+		});
+	}
+
+	// Note for this one we may need to include some logic to actually switch channels.
+	// for example if we're on screen 2, but we want to tune into screen 1 we can just switch.
+	setScreen = (screenId) => () => {
+		this.setState(() => {
+			return {screenId: screenId};
+		});
+		this.onToggle();
+	}
+
+
+	hideAdSpace = () => {
+		this.setState({adContent: '', showAd: false});
+	};
+
+	showAdSpace = ({adContent, duration}) => {
+		this.setState({adContent, showAd: true});
```

Comment:
Consider using `core/util.Job` to manage the timeout. I think it simplifies things a bit. For example, we should clear the current timeout here before starting another. We can do that with `if (this.adTimer) clearTimeout(this.addTimer)` but `Job` would do that for us when we called `startAfter(duration)`.